### PR TITLE
{2023.06}[foss/2023a] gmsh v4.12.2, basemap v1.39, geopandas v0.14.2

### PR DESCRIPTION
**Pull request by casparvl**

### CODE REVIEW

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2

**Comment by casparvl** [comment]
I'm confused... it fails with an FLTK download error. But... if I do `--force-download --fetch` on a local system, the URLs work fine... a `wget` also works fine. Let me retry, see if it was (two) hickup(s)

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2

**Comment by bedroge** [comment]
> I'm confused... it fails with an FLTK download error. But... if I do `--force-download --fetch` on a local system, the URLs work fine... a `wget` also works fine. Let me retry, see if it was (two) hickup(s)

It doesn't work on the login node either:
```
[bedroge@login1 19702]$ curl https://fltk.org/pub/fltk/1.3.8/fltk-1.3.8-source.tar.gz
curl: (7) Failed to connect to fltk.org port 443: Connection refused

[bedroge@login1 19702]$ /cvmfs/software.eessi.io/versions/2023.06/compat/linux/aarch64/usr/bin/wget https://fltk.org/pub/fltk/1.3.8/fltk-1.3.8-source.tar.gz
--2024-09-24 18:18:35--  https://fltk.org/pub/fltk/1.3.8/fltk-1.3.8-source.tar.gz
Resolving fltk.org... 173.230.155.139, 2600:3c01::f03c:91ff:fe9f:634b
Connecting to fltk.org|173.230.155.139|:443... failed: Connection refused.
Connecting to fltk.org|2600:3c01::f03c:91ff:fe9f:634b|:443... failed: Network is unreachable.
```

**Comment by bedroge** [comment]
I've manually downloaded and copied the source tarball to the bot account, so let's try again.

bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2

**Comment by bedroge** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen4
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by boegel** [comment]
@casparvl Failure for `aarch64/neoverse_*`?

**Comment by casparvl** [comment]
Thanks for the assist on the sources @bedroge . Just out of curiosity: I checked the config printed by EasyBuild in the log file. It didn't seem to look anywhere for sources other than the `/tmp` dir, but I guess that's not really true then? We have some persistent path in which it looks for sources?

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic

**Comment by casparvl** [comment]
Hm, I thought it was missing from the `eb --show-config` output. But maybe that was in a different stage, where that path wasn't configured properly or something. Anyway, good to know :)

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic

**Comment by casparvl** [comment]
`19849` fails with

```
[ 46%] Building CXX object src/TKXSBase/CMakeFiles/TKXSBase.dir/__/IFSelect/IFSelect_SignCategory.cxx.o
cd /tmp/bot/easybuild/build/occt/7.8.0/GCCcore-12.3.0/easybuild_obj/src/TKXSBase && /tmp/eb-6mvwws6j/eb-2h3t5bin/tmplyfnokml/rpath_wrappers/gxx_wrapper/g++ --sysroot=/cvmfs/software.eessi.io/versions/2023.06/compat/linux/aarch64 -DHAVE_FREEIMAGE -DHAVE_FREETYPE -DHAVE_OPENGL_EXT -DHAVE
_TK -DHAVE_XLIB -DOCC_CONVERT_SIGNALS -DTKXSBase_EXPORTS -I/cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/generic/software/Tcl/8.6.13-GCCcore-12.3.0/include -I/cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/generic/software/Tk/8.6.13-GCCcore-12.3.0/in
clude -I/cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/generic/software/freetype/2.13.0-GCCcore-12.3.0/include/freetype2 -I/cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/generic/software/FreeImage/3.18.0-GCCcore-12.3.0/include -I/tmp/bot/easybuild/bu
ild/occt/7.8.0/GCCcore-12.3.0/easybuild_obj/include/opencascade -O2 -ftree-vectorize -mcpu=generic -fno-math-errno -fexceptions -fPIC -Wall -Wextra -O3 -DNDEBUG -s -DNo_Exception -std=gnu++11 -fPIC -MD -MT src/TKXSBase/CMakeFiles/TKXSBase.dir/__/IFSelect/IFSelect_SignCategory.cxx.o -MF
 CMakeFiles/TKXSBase.dir/__/IFSelect/IFSelect_SignCategory.cxx.o.d -o CMakeFiles/TKXSBase.dir/__/IFSelect/IFSelect_SignCategory.cxx.o -c /tmp/bot/easybuild/build/occt/7.8.0/GCCcore-12.3.0/OCCT-7_8_0/src/IFSelect/IFSelect_SignCategory.cxx
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/aarch64/usr/bin/ld: /cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/generic/software/FreeImage/3.18.0-GCCcore-12.3.0/lib64/libfreeimage.so: undefined reference to `png_init_filter_functions_neon'
collect2: error: ld returned 1 exit status
make[2]: *** [src/DRAWEXE/CMakeFiles/DRAWEXE.dir/build.make:118: lin64/gcc/bin/DRAWEXE-7.8.0] Error 1
make[2]: Leaving directory '/tmp/bot/easybuild/build/occt/7.8.0/GCCcore-12.3.0/easybuild_obj'
make[1]: *** [CMakeFiles/Makefile2:3679: src/DRAWEXE/CMakeFiles/DRAWEXE.dir/all] Error 2
```
Failure occurs when building `occt-7.8.0`

**Comment by casparvl** [comment]
Interesting, I'm not getting the error when building with `EESSI-extend` on our local system. So... it might be due to the container (or actually, it might work on our system because it picks up OS stuff)

**Comment by bedroge** [comment]
> `19849` fails with
> 
> ```
> [ 46%] Building CXX object src/TKXSBase/CMakeFiles/TKXSBase.dir/__/IFSelect/IFSelect_SignCategory.cxx.o
> cd /tmp/bot/easybuild/build/occt/7.8.0/GCCcore-12.3.0/easybuild_obj/src/TKXSBase && /tmp/eb-6mvwws6j/eb-2h3t5bin/tmplyfnokml/rpath_wrappers/gxx_wrapper/g++ --sysroot=/cvmfs/software.eessi.io/versions/2023.06/compat/linux/aarch64 -DHAVE_FREEIMAGE -DHAVE_FREETYPE -DHAVE_OPENGL_EXT -DHAVE
> _TK -DHAVE_XLIB -DOCC_CONVERT_SIGNALS -DTKXSBase_EXPORTS -I/cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/generic/software/Tcl/8.6.13-GCCcore-12.3.0/include -I/cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/generic/software/Tk/8.6.13-GCCcore-12.3.0/in
> clude -I/cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/generic/software/freetype/2.13.0-GCCcore-12.3.0/include/freetype2 -I/cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/generic/software/FreeImage/3.18.0-GCCcore-12.3.0/include -I/tmp/bot/easybuild/bu
> ild/occt/7.8.0/GCCcore-12.3.0/easybuild_obj/include/opencascade -O2 -ftree-vectorize -mcpu=generic -fno-math-errno -fexceptions -fPIC -Wall -Wextra -O3 -DNDEBUG -s -DNo_Exception -std=gnu++11 -fPIC -MD -MT src/TKXSBase/CMakeFiles/TKXSBase.dir/__/IFSelect/IFSelect_SignCategory.cxx.o -MF
>  CMakeFiles/TKXSBase.dir/__/IFSelect/IFSelect_SignCategory.cxx.o.d -o CMakeFiles/TKXSBase.dir/__/IFSelect/IFSelect_SignCategory.cxx.o -c /tmp/bot/easybuild/build/occt/7.8.0/GCCcore-12.3.0/OCCT-7_8_0/src/IFSelect/IFSelect_SignCategory.cxx
> /cvmfs/software.eessi.io/versions/2023.06/compat/linux/aarch64/usr/bin/ld: /cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/generic/software/FreeImage/3.18.0-GCCcore-12.3.0/lib64/libfreeimage.so: undefined reference to `png_init_filter_functions_neon'
> collect2: error: ld returned 1 exit status
> make[2]: *** [src/DRAWEXE/CMakeFiles/DRAWEXE.dir/build.make:118: lin64/gcc/bin/DRAWEXE-7.8.0] Error 1
> make[2]: Leaving directory '/tmp/bot/easybuild/build/occt/7.8.0/GCCcore-12.3.0/easybuild_obj'
> make[1]: *** [CMakeFiles/Makefile2:3679: src/DRAWEXE/CMakeFiles/DRAWEXE.dir/all] Error 2
> ```
> 
> Failure occurs when building `occt-7.8.0`

https://github.com/NixOS/nixpkgs/issues/77653 and https://issues.guix.gnu.org/28261 are suggesting it's because of some old libpng version being shipped with FreeImage.

**Comment by casparvl** [comment]
Interesting, that undefined symbol should normally be provided by `libpng`. But, the README for `FreeImage` says that it ships its own versions https://github.com/WinMerge/freeimage/blob/master/README.linux . And indeed, it seems it does https://github.com/WinMerge/freeimage/tree/master/Source/LibPNG

The symbol is from here: https://github.com/WinMerge/freeimage/blob/78364ed94b4cb2d830805a26cb65bf5e6f523d58/Source/LibPNG/pngpriv.h#L142
https://github.com/WinMerge/freeimage/blob/78364ed94b4cb2d830805a26cb65bf5e6f523d58/Source/LibPNG/pngpriv.h#L2099

What I don't get, is where the _implementation_ of this function is then supposed to be. I don't see it in these sources. It does seem to be in to official libpng https://github.com/slembcke/libpng/blob/420d778366458d5273b1da95dda24cad1a3903c0/libpng/arm/arm_init.c#L57 But the FreeImage is quite explicit in saying you don't need the external deps. It's... weird...

**Comment by casparvl** [comment]
Ah, known issue, already since a long time https://sourceforge.net/p/freeimage/bugs/325/ and https://sourceforge.net/p/freeimage/discussion/36111/thread/cc4cd71c6e/?limit=25#7371 . Never fixed I guess. Now I also know why this didn't occur on my system: it's not ARM based :)

Workaround is to set `-DPNG_ARM_NEON_OPT=0` in `CFLAGS`.

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic

**Comment by bedroge** [comment]
> Ah, known issue, already since a long time https://sourceforge.net/p/freeimage/bugs/325/ and https://sourceforge.net/p/freeimage/discussion/36111/thread/cc4cd71c6e/?limit=25#7371 . Never fixed I guess. Now I also know why this didn't occur on my system: it's not ARM based :)
> 
> Workaround is to set `-DPNG_ARM_NEON_OPT=0` in `CFLAGS`.

I guess it's fine to use that for now, though I do wonder if we could somehow patch the FreeImage easyconfig in such a way that it uses a newer or patched libpng, or (even better) an EB-provided libpng.

**Comment by casparvl** [comment]
Well, I doubt that'll be easy. FreeImage _seems_ to be designed to use the internally shipped imaging libraries (libpng is one, but there are several others). So could you patch it? I guess so. But it might be pretty invasive and is pretty much opposing their own design vision. I'm not saying they are _right_ in shipping everything internally, but hey... it's not my application, not my decision ;-)

The real patch would be to pull in the missing part into the internal libpng. But that might also be a pretty major operation, we'll only know in hindsight.

Anyway, unless someone comes along with enough motivation and time to do that, let's leave it like this for now...

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen2
bot: build repo:eessi.io-2023.06-software arch:zen3
bot: build repo:eessi.io-2023.06-software arch:zen4
bot: build repo:eessi.io-2023.06-software arch:skylake_512
bot: build repo:eessi.io-2023.06-software arch:haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:neoverse_v1

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by bedroge** [comment]
> Well, I doubt that'll be easy. 

I thought so as well, but had a quick look at how Gentoo installs this, and maybe it's not that complex:

https://github.com/gentoo/gentoo/blob/master/media-libs/freeimage/freeimage-3.18.0-r11.ebuild#L63

**Comment by casparvl** [comment]
Do I understand correctly that they just throw away all those internally shipped libs, and use external deps?

**Comment by casparvl** [comment]
You want to implement it? ;-)

**Comment by bedroge** [comment]
> Do I understand correctly that they just throw away all those internally shipped libs, and use external deps?

Yes, I think so (and remove them from some Makefiles).


> You want to implement it? ;-)

For some reason, I was expecting such a question :-D

**Comment by bedroge** [comment]
But I had a quick look at it, and there's a huge patch that does the unbundling: https://dev.gentoo.org/~juippis/distfiles/tmp/freeimage-3.18.0-unbundling.patch). Porting that to EB is not trivial, since it also changes a lot of files before actually applying that patch. So I guess we just go ahead with the current workaround, and I/we can open an issue on the easyconfigs repo.

**Comment by casparvl** [comment]
Ok, I see that the last zen4 build didn't report back. I'll trigger it again. But then this can be merged, right?

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen4

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen4

**Comment by bedroge** [comment]
> So I guess we just go ahead with the current workaround, and I/we can open an issue on the easyconfigs repo.

https://github.com/easybuilders/easybuild-easyconfigs/issues/21487

File: easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.3-2023a.yml [hunk 1]
```diff
@@ -1,5 +1,8 @@
 easyconfigs:
+  - gmsh-4.12.2-foss-2023a.eb
+  - basemap-1.3.9-foss-2023a.eb
+  - geopandas-0.14.2-foss-2023a.eb
   - ccache-4.9-GCCcore-12.3.0.eb
   - GDB-13.2-GCCcore-12.3.0.eb
   - tmux-3.3a-GCCcore-12.3.0.eb
-  - Vim-9.1.0004-GCCcore-12.3.0.eb
+  - Vim-9.1.0004-GCCcore-12.3.0.eb
```
Comment (boegel) [inline easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.3-2023a.yml:8]: let's add new stuff at the end of the list:
```suggestion
  - ccache-4.9-GCCcore-12.3.0.eb
  - GDB-13.2-GCCcore-12.3.0.eb
  - tmux-3.3a-GCCcore-12.3.0.eb
  - Vim-9.1.0004-GCCcore-12.3.0.eb
  - gmsh-4.12.2-foss-2023a.eb
  - basemap-1.3.9-foss-2023a.eb
  - geopandas-0.14.2-foss-2023a.eb

**Review by boegel — CHANGES_REQUESTED** [review]

File: easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.3-2023a.yml [hunk 2]
```diff
@@ -1,5 +1,8 @@
 easyconfigs:
+  - gmsh-4.12.2-foss-2023a.eb
+  - basemap-1.3.9-foss-2023a.eb
+  - geopandas-0.14.2-foss-2023a.eb
```
Comment (boegel) [inline easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.3-2023a.yml:4]:
```suggestion
```

File: easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.3-2023a.yml (hunk 1 above)
Comment (boegel) [inline easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.3-2023a.yml:8]:
```suggestion
  - Vim-9.1.0004-GCCcore-12.3.0.eb
  - gmsh-4.12.2-foss-2023a.eb
  - basemap-1.3.9-foss-2023a.eb
  - geopandas-0.14.2-foss-2023a.eb
```

**Review by boegel — CHANGES_REQUESTED** [review]

**Comment by boegel** [comment]
all staging PRs merged...

**Comment by boegel** [comment]
@casparvl Missing installation check for `neoverse_v1` is failing due to problem in the hooks, can you look into fixing that?
```
  File "/home/runner/work/software-layer/software-layer/eb_hooks.py", line 78, in parse_hook
    PARSE_HOOKS[ec.name](ec, eprefix)
  File "/home/runner/work/software-layer/software-layer/eb_hooks.py", line 195, in parse_hook_casacore_disable_vectorize
    ec['toolchainopts']['vectorize'] = False
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
TypeError: 'NoneType' object does not support item assignment
```

This is a PITA, since the hooks are included in the ingested tarballs, so we'll need to trigger another build once fixed.

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen2
bot: build repo:eessi.io-2023.06-software arch:zen3
bot: build repo:eessi.io-2023.06-software arch:zen4
bot: build repo:eessi.io-2023.06-software arch:skylake_avx512
bot: build repo:eessi.io-2023.06-software arch:haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic
bot: build repo:eessi.io-2023.06-software arch:neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:neoverse_v1

**Comment by casparvl** [comment]
Ugh, it doesn't like that there's nothing to rebuild I guess. 

```
wrote file list to /tmp/tmp.3yobAKE2vW/files.list.txt
2023.06/init/easybuild/eb_hooks.py
wrote module file list to /tmp/tmp.3yobAKE2vW/module_files.list.txt
cp: cannot stat '/tmp/tmp.3yobAKE2vW/module_files.list.txt': No such file or directory
```
I.e. there's something in the `files.list.txt`, but not in the `module_files.list.txt`, I think that's the issue here.

Hmm, do we really never _just_ deploy new scripts, without new software?

**Comment by boegel** [comment]
> Ugh, it doesn't like that there's nothing to rebuild I guess.
> 
> ```
> wrote file list to /tmp/tmp.3yobAKE2vW/files.list.txt
> 2023.06/init/easybuild/eb_hooks.py
> wrote module file list to /tmp/tmp.3yobAKE2vW/module_files.list.txt
> cp: cannot stat '/tmp/tmp.3yobAKE2vW/module_files.list.txt': No such file or directory
> ```
> 
> I.e. there's something in the `files.list.txt`, but not in the `module_files.list.txt`, I think that's the issue here.
> 
> Hmm, do we really never _just_ deploy new scripts, without new software?

A PR that only touches scripts is supposed to work, and did work before, but anymore due to the changes in #673 .

I'll make a proposal to fix that...

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic

**Comment by boegel** [comment]
> > Ugh, it doesn't like that there's nothing to rebuild I guess.
> > ```
> > wrote file list to /tmp/tmp.3yobAKE2vW/files.list.txt
> > 2023.06/init/easybuild/eb_hooks.py
> > wrote module file list to /tmp/tmp.3yobAKE2vW/module_files.list.txt
> > cp: cannot stat '/tmp/tmp.3yobAKE2vW/module_files.list.txt': No such file or directory
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I.e. there's something in the `files.list.txt`, but not in the `module_files.list.txt`, I think that's the issue here.
> > Hmm, do we really never _just_ deploy new scripts, without new software?
> 
> A PR that only touches scripts is supposed to work, and did work before, but anymore due to the changes in #673 .
> 
> I'll make a proposal to fix that...

The changes in c9613e7 should fix the problem with creating the tarball.

@casparvl We should also update `test_suite.sh` so it takes into account that `module_files.list.txt` may not be there, but that's not critical for this PR.

**Comment by casparvl** [comment]
> @casparvl We should also update test_suite.sh so it takes into account that module_files.list.txt may not be there, but that's not critical for this PR.

I think the build step will still report as 'failed'. But I guess at least the tarball _will_ be created correctly. So if you're bold enough to deploy even when the bot calls the build step 'failed', we're good ;-)

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen4
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1

**Comment by casparvl** [comment]
no wait, that's not true. The check-build is done before testing, so those failures are not in the logs yet. Ok, we should be good then :)

I'll create a fix for the other issue.

**Comment by boegel** [comment]
@casparvl Just a single rebuild was good enough, since `eb_hooks.py` is not in a CPU-specific path :)

So I've merged one staging PR, I'll close the others...

**Review by boegel — APPROVED** [review]

**Pull request merged**